src/Security/MemberAuthenticator/LoginHandler.php

```diff
@@ -0,0 +1,254 @@
+use SilverStripe\Control\HTTPResponse;
+use SilverStripe\Control\Session;
+use SilverStripe\Control\RequestHandler;
+use SilverStripe\ORM\ValidationResult;
+use SilverStripe\Security\Security;
+use SilverStripe\Security\Member;
+use SilverStripe\Core\Injector\Injector;
+use SilverStripe\Security\IdentityStore;
+
+/**
+ * Handle login requests from MemberLoginForm
+ */
+class LoginHandler extends RequestHandler
+{
+    /**
+     * @var Authenticator
+     */
+    protected $authenticator;
+
+    private static $url_handlers = [
+        '' => 'login',
+    ];
+
+    /**
+     * Since the logout and dologin actions may be conditionally removed, it's necessary to ensure these
+     * remain valid actions regardless of the member login state.
+     *
+     * @var array
+     * @config
+     */
+    private static $allowed_actions = [
+        'login',
+        'LoginForm',
+        'logout',
+    ];
+
+    private $link;
+
+    /**
+     * @param string $link The URL to recreate this request handler
+     * @param Authenticator $authenticator The
+     */
+    public function __construct($link, Authenticator $authenticator)
+    {
+        $this->link = $link;
+        $this->authenticator = $authenticator;
+        parent::__construct($link, $this);
+    }
+
+    /**
+     * Return a link to this request handler.
+     * The link returned is supplied in the constructor
+     * @param null $action
+     * @return string
+     */
+    public function link($action = null)
+    {
+        if ($action) {
+            return Controller::join_links($this->link, $action);
+        }
+
+        return $this->link;
+    }
+
+    /**
+     * URL handler for the log-in screen
+     */
+    public function login()
+    {
+        return [
+            'Form' => $this->loginForm(),
+        ];
+    }
+
+    /**
+     * Return the MemberLoginForm form
+     */
+    public function loginForm()
+    {
+        return LoginForm::create(
+            $this,
+            get_class($this->authenticator),
+            'LoginForm'
+        );
+    }
+
+    /**
+     * Login form handler method
+     *
+     * This method is called when the user finishes the login flow
+     *
+     * @param array $data Submitted data
+     * @param LoginForm $form
+     * @return HTTPResponse
+     */
+    public function doLogin($data, $form)
+    {
+        $failureMessage = null;
+
+        // Successful login
+        if ($member = $this->checkLogin($data, $failureMessage)) {
+            $this->performLogin($member, $data, $form->getRequestHandler()->getRequest());
+
+            return $this->redirectAfterSuccessfulLogin();
+        }
+
+        $form->sessionMessage($failureMessage, 'bad');
+
+        // Failed login
+
+        /** @skipUpgrade */
+        if (array_key_exists('Email', $data)) {
+            Session::set('SessionForms.MemberLoginForm.Email', $data['Email']);
+            Session::set('SessionForms.MemberLoginForm.Remember', isset($data['Remember']));
```

Comment:
Good call, we might want to separate those checks out into a single method that is publicly available to check if the Remember flag is usable or always false?